packages/kbn-health-gateway-server/src/kibana/routes/root.ts

```diff
@@ -6,164 +6,161 @@
-    statusCode: hasUnhealthyResponse ? 503 : 200,
-  };
-}
+  async handler(request: Request, toolkit: ResponseToolkit): Promise<ResponseObject> {
+    const body = await this.poll();
+    const code = RootRoute.STATUS_CODE[body.status];
 
-function isUnhealthyResponse(response: PromiseFulfilledResult<Response> | PromiseRejectedResult) {
-  return (
-    response.status === 'rejected' || !(isSuccess(response.value) || isUnauthorized(response.value))
-  );
-}
+    this.logger.debug(`Returning ${code} response with body: ${JSON.stringify(body)}`);
 
-function isUnauthorized({ status, headers }: Response): boolean {
-  return status === 401 && headers.has('www-authenticate');
-}
+    return toolkit.response(body).type('application/json').code(code);
+  }
 
-function isSuccess({ status }: Response): boolean {
-  return (status >= 200 && status <= 299) || status === 302;
-}
+  private async poll(): Promise<RootRouteResponse> {
+    const hosts = await Promise.all(this.kibanaConfig.hosts.map(this.pollHost.bind(this)));
+    const statuses = chain(hosts).map('status').uniq().value();
+    const status = statuses.length <= 1 ? statuses[0] ?? 'healthy' : 'unhealthy';
 
-function generateAgentConfig(sslConfig: KibanaConfig['ssl']) {
-  const options: https.AgentOptions = {
-    ca: sslConfig.certificateAuthorities,
-    cert: sslConfig.certificate,
-  };
-
-  const verificationMode = sslConfig.verificationMode;
-  switch (verificationMode) {
-    case 'none':
-      options.rejectUnauthorized = false;
-      break;
-    case 'certificate':
-      options.rejectUnauthorized = true;
-      // by default, NodeJS is checking the server identify
-      options.checkServerIdentity = () => undefined;
-      break;
-    case 'full':
-      options.rejectUnauthorized = true;
-      break;
-    default:
-      throw new Error(`Unknown ssl verificationMode: ${verificationMode}`);
+    return {
+      status,
+      hosts,
+    };
   }
 
-  return options;
-}
+  private async pollHost(host: string): Promise<HostStatus> {
+    const url = `${host}${RootRoute.POLL_ROUTE}`;
+    this.logger.debug(`Requesting ${url}`);
 
-function configureFetch(kibanaConfig: KibanaConfig) {
-  let agent: https.Agent;
+    try {
+      const response = await this.fetch(url);
+      const status = RootRoute.isHealthy(response) ? 'healthy' : 'unhealthy';
+      this.logger.debug(`${capitalize(status)} response from ${url} with code ${response.status}`);
+
+      return {
+        host,
+        status,
+        code: response.status,
+      };
+    } catch (error) {
+      this.logger.error(error);
+
+      if (error.name === 'AbortError') {
+        this.logger.error(`Request timeout for ${url}`);
+
+        return {
+          host,
+          status: 'timeout',
+          message: error.message,
+        };
+      }
+
+      this.logger.error(`Failed response from ${url}: ${error.message}`);
+
+      return {
+        host,
+        status: 'failure',
+        message: error.message,
```

Comment:
I think it's safer not to return raw exception messages. E.g. if a plugin's route handler throws Kibana will return `500` without the details of the exception and users need to look at the server logs for full details.

So if we know what the problem was we could say e.g. `message: "Request timeout when trying to connect to host"` and for the general case we probably need to just keep a generic message like "Error while trying to get host health".

Reply:
I've removed that entirely, as we have a dedicated status code for the timeout errors. Otherwise, it would be redundant as it is always the same per status type.